extensions/data-transfer/portability-data-transfer-generic/src/main/java/org/datatransferproject/datatransfer/generic/GenericImporter.java

```diff
@@ -0,0 +1,172 @@
+      return error;
+    }
+
+    public Optional<String> getErrorDescription() {
+      return errorDescription;
+    }
+
+    public String toString() {
+      StringBuilder builder = new StringBuilder();
+      builder.append(error);
+      if (errorDescription.isPresent()) {
+        builder.append(" - ");
+        builder.append(errorDescription.get());
+      }
+      return builder.toString();
+    }
+  }
+
+  ContainerSerializer<C, R> containerSerializer;
+  URL endpoint;
+  Monitor monitor;
+  AppCredentials appCredentials;
+  OkHttpClient client = new OkHttpClient();
+  ObjectMapper om = new ObjectMapper();
+  Map<UUID, OAuthTokenManager> jobTokenManagerMap = new HashMap<>();
+
+  static final MediaType JSON = MediaType.parse("application/json");
+
+  public GenericImporter(
+      ContainerSerializer<C, R> containerSerializer,
+      AppCredentials appCredentials,
+      URL endpoint,
+      Monitor monitor) {
+    this.monitor = monitor;
+    this.appCredentials = appCredentials;
+    this.endpoint = endpoint;
+    this.containerSerializer = containerSerializer;
+    configureObjectMapper(om);
+  }
+
+  @VisibleForTesting
+  static void configureObjectMapper(ObjectMapper objectMapper) {
+    // ZonedDateTime and friends
+    objectMapper.registerModule(new JavaTimeModule());
+    // Optional fields
+    objectMapper.registerModule(new Jdk8Module());
+    // ISO timestamps rather than unix epoch seconds
+    objectMapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
+  }
+
+  @Override
+  public ImportResult importItem(
+      UUID jobId,
+      IdempotentImportExecutor idempotentExecutor,
+      TokensAndUrlAuthData initialAuthData,
+      C data)
+      throws Exception {
+    OAuthTokenManager tokenManager =
+        jobTokenManagerMap.computeIfAbsent(
+            jobId,
+            ignored -> new OAuthTokenManager(initialAuthData, appCredentials, client, monitor));
+    for (ImportableData<R> importableData : containerSerializer.apply(data)) {
+      idempotentExecutor.executeAndSwallowIOExceptions(
+          importableData.getIdempotentId(),
+          importableData.getName(),
+          () ->
+              tokenManager.withAuthData(
+                  authData -> importSingleItem(jobId, authData, importableData)));
+    }
+    return new ImportResult(ResultType.OK);
+  }
+
+  boolean parseResponse(Response response) throws IOException, InvalidTokenException {
+    if (response.code() >= 400) {
+      byte[] body = response.body().bytes();
+      ErrorResponse error;
+      try {
+        error = om.readValue(body, ErrorResponse.class);
+      } catch (JsonParseException | JsonMappingException e) {
+        throw new IOException(
+            format(
+                "Unexpected response (%d) '%s'",
+                response.code(), new String(body, StandardCharsets.UTF_8)),
+            e);
+      }
+
+      if (response.code() == 401 && error.getError().equals("invalid_token")) {
```

Comment:
isn't 401 already mean invalid token one way or the other ? I feel like the additional  check for error message is redundant.

Reply:
It would probably be safe to refresh in most cases, but there are other reasons to send 401, e.g. missing token, invalid token format. In those cases the following token refresh request is likely to fail, which could mask the underlying issue slightly when debugging. Very minor, and I think you're right that it's likely the only reason for a 401 in most implementations, but this is one area I wanted the behaviour to be specifically defined

See https://github.com/dtinit/data-transfer-project/pull/1413 where this was fixed recently for the Microsoft extension which was just checking 401 - maybe @jzacsh can weigh in